### PR TITLE
Pyface Color Class and Trait Type

### LINKED ISSUE #476: Basic Color and Font traits in Pyface
We are using the `Color` and `Font` traits from `traits.api`.  These implicitly depend on TraitsUI, and will be deprecated in Traits.

We should refactor so that the toolkit definitions of the basic Traits (possibly without editors) are defined in Pyface, but imported into TraitsUI which have the TraitFactory versions which should be  used by most end-users.

### PR DESCRIPTION
**Pull request by corranwebster**
This builds on #520 to create a `Color` class and corresponding `PyfaceColor` trait to represent a color in a toolkit independent way.  It also includes a `ColorDialog` class built on top of these classes.

Together with #520 this resolves #476.

This should not be merged until #520 is merged.  This requires features from Traits 6.1.

### CODE REVIEW

File: pyface/tests/test_color.py [hunk 1]
```diff
@@ -0,0 +1,468 @@
...
+
+    def test_hex_6(self):
+        space, channels = parse_hex('#06c')
+        self.assertEqual(space, 'rgb')
+        self.assertEqual(channels, (0.0, 0.4, 0.8))
+
+    def test_hex_8(self):
+        space, channels = parse_hex('#06cf')
+        self.assertEqual(space, 'rgba')
+        self.assertEqual(channels, (0.0, 0.4, 0.8, 1.0))
```
Comment (ievacerny) [inline pyface/tests/test_color.py:179]:
```suggestion
    def test_hex_6(self):
        space, channels = parse_hex('#0066cc')
        self.assertEqual(space, 'rgb')
        self.assertEqual(channels, (0.0, 0.4, 0.8))

    def test_hex_8(self):
        space, channels = parse_hex('#0066ccff')
        self.assertEqual(space, 'rgba')
        self.assertEqual(channels, (0.0, 0.4, 0.8, 1.0))
```

File: pyface/ui/null/color.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    Returns
+    -------
+    rgba : tuple
+        A tuple of 4 floating point values between 0.0 and 1.0 inclusive.
+    """
+    return ints_to_channels(color)
+
+
+def rgba_to_toolkit_color(rgba):
+    """ Convert a hex tuple to an RGBA tuple.
```
Comment (ievacerny) [inline pyface/ui/null/color.py:39]: Copy-paste docstring

File: pyface/color_dialog.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+"""
+
+from .constant import OK
+from .toolkit import toolkit_object
+
+
+ColorDialog = toolkit_object("color_dialog:ColorDialog")
+
+
+def get_color(parent, color, show_alpha=False):
```
Comment (ievacerny) [inline pyface/color_dialog.py:21]: Similarly to `get_font` - should `parent` and `show_alpha` be passed to the actual dialog? And there are no tests for this function

File: pyface/tests/test_color.py [hunk 4]
```diff
@@ -0,0 +1,468 @@
...
+        self.assertEqual(color.rgba, (0.4, 0.2, 0.6, 1.0))
+
+    def test_get_rgb(self):
+        color = Color(rgba=(0.4, 0.2, 0.6, 0.8))
+        self.assertEqual(color.rgb, (0.4, 0.2, 0.6))
+
+    def test_set_rgb(self):
+        color = Color(rgba=(0.4, 0.2, 0.6, 0.8))
+        color.rgb = (0.6, 0.8, 0.4)
+        self.assertEqual(color.rgba, (0.6, 0.8, 0.4, 0.8))
```
Comment (ievacerny) [inline pyface/tests/test_color.py:371]: Missing tests for `hsva`, `hsv`, `hlsa` and `hls` properties

File: pyface/ui/wx/color_dialog.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        self.color = Color.from_toolkit(wx_color)
+        super(ColorDialog, self).close()
+
+    # ------------------------------------------------------------------------
+    # 'IWidget' interface.
+    # ------------------------------------------------------------------------
+
+    def _create_control(self, parent):
+        wx_color = self.color.to_toolkit()
+        data = wx.ColorData()
```
Comment (ievacerny) [inline pyface/ui/wx/color_dialog.py:59]: Trying to open a dialog in wx fails with error:
```
>>> get_color(None, "red")

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/icernyte/Documents/repos/pyface/pyface/color_dialog.py", line 39, in get_color
    result = dialog.open()
  File "/Users/icernyte/Documents/repos/pyface/pyface/i_dialog.py", line 164, in open
    self._create()
  File "/Users/icernyte/Documents/repos/pyface/pyface/i_dialog.py", line 183, in _create
    super(MDialog, self)._create()
  File "/Users/icernyte/Documents/repos/pyface/pyface/i_widget.py", line 108, in _create
    self.control = self._create_control(self.parent)
  File "/Users/icernyte/Documents/repos/pyface/pyface/ui/wx/color_dialog.py", line 59, in _create_control
    data = wx.ColorData()
AttributeError: module 'wx' has no attribute 'ColorData'
```

Comment (corranwebster) [reply pyface/ui/wx/color_dialog.py:59]: WxPython uses UK spellings!

**Comment by corranwebster** [comment]
Pending CI passing, this should be ready for re-review.

File: pyface/ui/wx/color_dialog.py (hunk 5 above)
Comment (ievacerny) [reply pyface/ui/wx/color_dialog.py]: Opening the dialog still fails, but now with a different error on the next line:
```
>>> get_color(None, "red")

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/icernyte/Documents/repos/pyface/pyface/color_dialog.py", line 39, in get_color
    result = dialog.open()
  File "/Users/icernyte/Documents/repos/pyface/pyface/i_dialog.py", line 164, in open
    self._create()
  File "/Users/icernyte/Documents/repos/pyface/pyface/i_dialog.py", line 183, in _create
    super(MDialog, self)._create()
  File "/Users/icernyte/Documents/repos/pyface/pyface/i_widget.py", line 108, in _create
    self.control = self._create_control(self.parent)
  File "/Users/icernyte/Documents/repos/pyface/pyface/ui/wx/color_dialog.py", line 60, in _create_control
    data.SetInitialColour(wx_colour)
AttributeError: 'ColourData' object has no attribute 'SetInitialColour'
```

**Comment by corranwebster** [comment]
I think this is ready for one more go-around with review.

File: pyface/color_dialog.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+    Returns
+    -------
+    color : Color or None
+        The selected color, or None if the user made no selection.
+    """
+    dialog = ColorDialog(parent=parent, color=color, show_alpha=show_alpha)
+    result = dialog.open()
+    if result == OK:
+        return dialog.color
```
Comment (ievacerny) [inline pyface/color_dialog.py:41]: Closing the dialog via system's "x" button returns the color that was passed to this function when using wx. With qt it returns `None` as expected.

**Comment by corranwebster** [comment]
Everything in here has now been pulled out into a separate PR.

**Pull request closed without merging**